dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/loadbalance/RoundRobinLoadBalance.java

```diff
@@ -17,68 +17,147 @@
-                } else {
-                    currentWeight = sequence.get() % maxWeight;
+        synchronized (map) {
+            int totalWeight = 0;
+            int maxCurrent = Integer.MIN_VALUE;
+            long now = System.currentTimeMillis();
+            boolean needRecycle = false;
+            Invoker<T> selectedInvoker = null;
+            WeightedRoundRobin selectedWRR = null;
+            for (Invoker<T> invoker : invokers) {
+                String identifyString = invoker.getUrl().toIdentityString();
+                WeightedRoundRobin weightedRoundRobin = map.get(identifyString);
+                int weight = getWeight(invoker, invocation);
+                if (weight < 0) {
+                    weight = 0;
+                }
+                if (weightedRoundRobin == null) {
+                    weightedRoundRobin = new WeightedRoundRobin();
+                    weightedRoundRobin.setCurrent(0);
+                    weightedRoundRobin.setWeight(weight);
+                    weightedRoundRobin.setLastRecycle(now);
+                    map.put(identifyString, weightedRoundRobin);
                 }
-                if (getWeight(nonZeroWeightedInvokers.get(index), invocation) > currentWeight) {
-                    return nonZeroWeightedInvokers.get(index);
+                if (weight != weightedRoundRobin.getWeight()) {
+                    //weight changed
+                    weightedRoundRobin.setCurrent(0);
+                    weightedRoundRobin.setWeight(weight);
                 }
+                int cur = weightedRoundRobin.increaseWeight();
+                weightedRoundRobin.setLastUpdate(now);
+                if (!needRecycle && now - weightedRoundRobin.getLastRecycle() > RECYCLE_PERIOD) {
+                    //try to recycle useless item every 60s if there is one outdated
+                    needRecycle = true;
+                }
+                if (cur > maxCurrent) {
+                    maxCurrent = cur;
+                    selectedInvoker = invoker;
+                    selectedWRR = weightedRoundRobin;
+                }
+                totalWeight += weight;
+            }
+            //recycle
+            if (needRecycle) {
+                Iterator<Entry<String, WeightedRoundRobin>> it = map.entrySet().iterator();
+                while (it.hasNext()) {
+                    Entry<String, WeightedRoundRobin> item = it.next();
+                    if (now - item.getValue().getLastUpdate() > RECYCLE_PERIOD) {
+                        it.remove();
+                    } else {
+                        //pass
+                        item.getValue().setLastRecycle(now);
+                    }
+                }
+            }
+            if (selectedInvoker != null) {
+                selectedWRR.sel(totalWeight);
+                return selectedInvoker;
             }
         }
-        // Round robin
-        return invokers.get(sequence.getAndIncrement() % length);
+        return null;

```

Comment:
one more thing, should we fall back to `random` here?

Reply:
It should not happen. So return the `first` item or do a `simple random` selecting without weight?